flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java

```diff
@@ -0,0 +1,155 @@
+	 * {@link Types#SQL_TIMESTAMP} field into the rowtime attribute.
+	 *
+	 * @param fieldName The field to convert into a rowtime attribute.
+	 */
+	public Rowtime timestampsFromField(String fieldName) {
+		internalProperties.putString(
+			ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD);
+		internalProperties.putString(ROWTIME_TIMESTAMPS_FROM, fieldName);
+		return this;
+	}
+
+	/**
+	 * Sets a built-in timestamp extractor that converts the assigned timestamps from
+	 * a DataStream API record into the rowtime attribute and thus preserves the assigned
+	 * timestamps from the source.
+	 *
+	 * <p>Note: This extractor only works in streaming environments.
+	 */
+	public Rowtime timestampsFromSource() {
+		internalProperties.putString(
+			ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_SOURCE);
+		return this;
+	}
+
+	/**
+	 * Sets a custom timestamp extractor to be used for the rowtime attribute.
+	 *
+	 * @param extractor The {@link TimestampExtractor} to extract the rowtime attribute
+	 *                  from the physical type.
+	 */
+	public Rowtime timestampsFromExtractor(TimestampExtractor extractor) {
+		internalProperties.putString(ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM);
+		internalProperties.putString(ROWTIME_TIMESTAMPS_CLASS, extractor.getClass().getName());
+		internalProperties.putString(ROWTIME_TIMESTAMPS_SERIALIZED, EncodingUtils.encodeObjectToString(extractor));
+		return this;
+	}
+
+	/**
+	 * Sets a built-in watermark strategy for ascending rowtime attributes.
+	 *
+	 * <p>Emits a watermark of the maximum observed timestamp so far minus 1.
+	 * Rows that have a timestamp equal to the max timestamp are not late.
+	 */
+	public Rowtime watermarksPeriodicAscending() {
+		internalProperties.putString(
+			ROWTIME_WATERMARKS_TYPE, ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_ASCENDING);
+		return this;
+	}
+
+	/**
+	 * Sets a built-in watermark strategy for rowtime attributes which are out-of-order by a bounded
+	 * time interval.
+	 *
+	 * <p>Emits watermarks which are the maximum observed timestamp minus the specified delay.
+	 *
+	 * @param delay delay in milliseconds
+	 */
+	public Rowtime watermarksPeriodicBounded(long delay) {
+		internalProperties.putString(
+			ROWTIME_WATERMARKS_TYPE, ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_BOUNDED);
+		internalProperties.putLong(ROWTIME_WATERMARKS_DELAY, delay);
+		return this;
+	}
+
+	/**
+	 * Sets a built-in watermark strategy which indicates the watermarks should be preserved from
+	 * the underlying DataStream API and thus preserves the assigned watermarks from the source.
+	 */
+	public Rowtime watermarksFromSource() {
+		internalProperties.putString(
+			ROWTIME_WATERMARKS_TYPE, ROWTIME_WATERMARKS_TYPE_VALUE_FROM_SOURCE);
+		return this;
+	}
+
+	/**
+	 * Sets a custom watermark strategy to be used for the rowtime attribute.
+	 */
+	public Rowtime watermarksFromStrategy(WatermarkStrategy strategy) {
+		internalProperties.putString(
```

Comment:
This changes the original logic. Before we were only serializing as the last option. Serializing is dangerous as we might not be able to restore the base64 string. Please reintroduce the original logic.